### ISSUE TITLE
GPU-native preprocessing for `torch.Tensor` inputs

### ISSUE THREAD

**Issue by David-rn** [ultralytics/ultralytics#24756]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

#  Preprocessing for `torch.Tensor` inputs

## Context

When using GPU-based video pipelines (GStreamer + DeepStream, FFmpeg with CUDA, etc.), decoded frames are available directly as GPU tensors with shape `(B, C, H_orig, W_orig)` and values in `[0, 255]`. As far as I can tell, passing a `torch.Tensor` to the predictor currently requires it to already be in `(B, C, 640, 640)` shape and normalized to `[0, 1]`. That means preprocessing has to happen outside of Ultralytics to then feed the correct tensor.

## Question

Is there a way to have Ultralytics apply the preprocessing stage (letterbox + resize + normalization) even when the input is a `torch.Tensor`, keeping everything on-device?

Looking at `BasePredictor.preprocess()` in [`engine/predictor.py`](https://github.com/ultralytics/ultralytics/blob/ec8a844f5a7fea50e339fc7230b85cd963c3c668/ultralytics/engine/predictor.py#L154), the `pre_transform()` call and `/= 255` normalization are both gated behind `not_tensor`, so tensor inputs bypass them entirely:

```python
not_tensor = not isinstance(im, torch.Tensor)
if not_tensor:
    im = np.stack(self.pre_transform(im)) 
    ...
if not_tensor:
    im /= 255
```

I couldn't find a flag or override to enable preprocessing for tensor inputs so I am not sure if I'm missing something.  For pipelines where frames never leave the GPU, a GPU-native preprocessing path would avoid unnecessary device transfers and keep the full pipeline on-device. 

Happy to contribute a PR if this is considered a worthwhile addition. 

Many thanks in advance!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @David-rn, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue triaged quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Since this appears to be a usage/question or possible feature-related behavior around preprocessing for `torch.Tensor` inputs, thank you for including context and code references 🙌 If you believe this is also a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello.

> Is there a way to have Ultralytics apply the preprocessing stage (letterbox + resize + normalization) even when the input is a torch.Tensor, keeping everything on-device?

There isn't a way to enable this without modifying source code which would be pretty much the same as just preprocessing it manually outside. Ultralytics preprocessing expects numpy array, so it won't be able to handle tensor preprocessing.

**Comment by David-rn** [comment]
Hello!

Thank you for the response! Okay, no problem, just wanted to make sure that I wasn't missing any standard preprocessing utility that the library had already implemented for such use cases

**Comment by glenn-jocher** [comment]
Correct, there’s no built-in `torch.Tensor` letterbox/normalize path in `predict()` today; for fully on-device pipelines the recommended approach is external GPU preprocessing, and the new [NVIDIA DALI guide](https://docs.ultralytics.com/guides/nvidia-dali) shows how to do that before passing a BCHW `[0, 1]` tensor to `model.predict()`.

**Comment by David-rn** [comment]
Perfect! I wasn't aware of the NVIDIA DALI guide. Many thanks!!

**Comment by glenn-jocher** [comment]
Glad that helped; if you decide to prototype a GPU-native `torch.Tensor` preprocessing path, a focused PR that preserves current `predict()` behavior by default would be welcome for review.

**Comment by David-rn** [comment]
Sure! I'll give it a try and keep you posted.

**Comment by glenn-jocher** [comment]
Sounds good — if you open a PR, please keep it narrowly scoped to tensor preprocessing, avoid adding new `default.yaml` args if possible, and include a small test plus a docs note so we can review it cleanly.
